File: poc/src/features/employee/api/getEmployeePositions.ts
export interface EmployeePositionItem {
  projectId: string;
  projectName: string;
  contractId: string;
  contractName: string;
  position: string;
  startDate: string;
  endDate: string | null;
}

export interface GetEmployeePositionsResponse {
  positions: EmployeePositionItem[];
}

const mockResponse: GetEmployeePositionsResponse = {
  positions: [
    {
      projectId: "11111111-1111-1111-1111-111111111111",
      projectName: "Název projektu",
      contractId: "22222222-2222-2222-2222-222222222222",
      contractName: "Název zakázky",
      position: "01-01 - název pozice",
      startDate: "2024-03-01",
      endDate: "2025-10-31",
    },
    {
      projectId: "11111111-1111-1111-1111-111111111111",
      projectName: "Název projektu",
      contractId: "33333333-3333-3333-3333-333333333333",
      contractName: "Název zakázky",
      position: "01-01 - název pozice",
      startDate: "2024-10-01",
      endDate: "2025-11-01",
    },
    {
      projectId: "44444444-4444-4444-4444-444444444444",
      projectName: "Název projektu",
      contractId: "55555555-5555-5555-5555-555555555555",
      contractName: "Název zakázky",
      position: "01-01 - název pozice",
      startDate: "2023-07-15",
      endDate: "2025-12-31",
    },
    {
      projectId: "44444444-4444-4444-4444-444444444444",
      projectName: "Název projektu",
      contractId: "66666666-6666-6666-6666-666666666666",
      contractName: "Název zakázky",
      position: "01-01 - název pozice",
      startDate: "2023-01-01",
      endDate: "2023-06-30",
    },
  ],
};

export const getEmployeePositions = (_employeeId: string) => {
  return {
    promise: (async (): Promise<GetEmployeePositionsResponse> => {
      await new Promise((resolve) => setTimeout(resolve, 1200));
      return mockResponse;
    })(),
  };
};

File: backend/Timesheets.Api/Employees/Endpoints/GetEmployeePositions.cs
using Microsoft.AspNetCore.Http.HttpResults;
using Microsoft.EntityFrameworkCore;
using Timesheets.Api.Auth;
using Timesheets.Api.Data;

namespace Timesheets.Api.Employees.Endpoints;

public sealed class GetEmployeePositions : IEndpoint
{
    public static void Map(IEndpointRouteBuilder app) =>
        app.MapGet("/{id}/positions", Handle)
           .WithSummary("Get Employee Positions");

    public sealed record EmployeePositionItem(
        Guid Id,
        Guid ProjectId,
        string ProjectName,
        DateTime ProjectStartDate,
        DateTime? ProjectEndDate,
        Guid ContractId,
        string ContractRegistrationNumber,
        string PositionCode,
        string Position,
        decimal Workload,
        DateTime StartDate,
        DateTime? EndDate
    );
    public sealed record Response(Guid EmployeeId, IEnumerable<EmployeePositionItem> Positions);

    private static async Task<Results<Ok<Response>, NotFound, ForbidHttpResult>> Handle(Guid id, AppDbContext dbContext, ICurrentUser user, CancellationToken cancellationToken)
    {
        if (!await user.CanAccessEmployeeAsync(id, cancellationToken))
        {
            return TypedResults.Forbid();
        }

        bool employeeExists = await dbContext.Employees
            .AsNoTracking()
            .AnyAsync(e => e.Id == id, cancellationToken);

        if (!employeeExists)
        {
            return TypedResults.NotFound();
        }

        List<EmployeePositionItem> positions = await dbContext.ContractEmployees
            .AsNoTracking()
            .Where(e => e.EmployeeId == id)
            .OrderBy(e => e.ContractPosition.Contract.Project.Name)
            .ThenBy(e => e.ContractPosition.Contract.RegistrationNumber)
            .ThenBy(e => e.StartDate)
            .Select(e => new EmployeePositionItem(
                e.Id,
                e.ContractPosition.Contract.Project.Id,
                e.ContractPosition.Contract.Project.Name,
                e.ContractPosition.Contract.Project.StartDate,
                e.ContractPosition.Contract.Project.EndDate,
                e.ContractPosition.Contract.Id,
                e.ContractPosition.Contract.RegistrationNumber,
                e.ContractPosition.Code,
                e.ContractPosition.Name,
                e.Workload,
                e.StartDate,
                e.EndDate
            ))
            .ToListAsync(cancellationToken);

        return TypedResults.Ok(new Response(id, positions));
    }
}
